fw/sock_clnt.c

```diff
@@ -181,39 +181,53 @@ tfw_h2_sk_prepare_xmit(struct sock *sk, struct sk_buff *skb, unsigned int mss_no
 	/* Avoid adding data to linear part of skb. */
 	skb->tail_lock = 1;
 
-#define FRAME_HEADERS_SHOULD_BE_MADE(flags)				\
-	((flags & SS_F_HTTT2_FRAME_HEADERS)				\
-	 && !(flags & SS_F_HTTT2_FRAME_HEADERS_DONE))
+#define FRAME_HEADERS_SHOULD_BE_MADE(flags)					\
+	(flags & SS_F_HTTT2_FRAME_HEADERS)
 
-#define FRAME_DATA_SHOULD_BE_MADE(flags)				\
-	((flags & SS_F_HTTT2_FRAME_DATA)				\
-	 && !(flags & SS_F_HTTT2_FRAME_DATA_DONE))
+#define FRAME_DATA_SHOULD_BE_MADE(flags)					\
+	(flags & SS_F_HTTT2_FRAME_DATA)
 
-#define FRAME_HEADERS_OR_DATA_SHOULD_BE_MADE(flags)			\
-	(FRAME_HEADERS_SHOULD_BE_MADE(flags)				\
+#define FRAME_HEADERS_OR_DATA_SHOULD_BE_MADE(flags)				\
+	(FRAME_HEADERS_SHOULD_BE_MADE(flags)					\
 	 || FRAME_DATA_SHOULD_BE_MADE(flags))
 
-#define FRAME_ALREADY_PREPARED(flags)					\
+#define FRAME_ALREADY_PREPARED(flags)						\
 	(flags & SS_F_HTTP2_FRAME_PREPARED)
 
-	if (FRAME_HEADERS_OR_DATA_SHOULD_BE_MADE(flags)) {
-		if (!stream) {
-			T_WARN("%s: stream with id (%u) already closed",
-			       __func__, skb_priv);
-			/*
-			 * We don't purge tcp queue, because we can still
-			 * send data for other streams.
-			 */
-			r = -EPIPE;
+#define FRAME_ALREADY_DONE(flags)						\
+	((flags & SS_F_HTTT2_FRAME_HEADERS_DONE)				\
+	 || (flags & SS_F_HTTT2_FRAME_DATA_DONE))
+
+#define CHECK_STREAM_IS_PRESENT(stream)						\
+	if (!stream) {								\
+			T_WARN("%s: stream with id (%u) already closed",	\
+			       __func__, skb_priv);				\
+			/*							\
+			 * We don't purge tcp queue, because we can still	\
+			 * send data for other streams.				\
+			 */							\
+			r = -EPIPE;						\
+			goto ret;						\
+	}
+
+	BUG_ON(FRAME_ALREADY_PREPARED(flags));
+
+	if (FRAME_ALREADY_DONE(flags)) {
```

Comment:
We discussed this place on the call - could you please write a good comment that we need the code to process an skb, which we already seen in the code between `tcp_tfw_sk_prepare_xmit()` and `tcp_tfw_sk_write_xmit()` calls

Reply:
Yes fixed